### ISSUE TITLE
Slices don't implement copy.

### ISSUE THREAD

**Issue by mahkoh** [rust-lang/rust#30483]
``` rust
[T]: Copy where T: Copy
```

does not hold even though it could. Together with `size_of_val`, this would allow one to write various functions generic over all Copy types whether they are sized or unsized.

**Comment by mahkoh** [comment]
It's seems like Copy requires Sized but i don't see a reason for that either.

**Comment by eddyb** [comment]
See #25776.

**Comment by mahkoh** [comment]
The case where I noticed this is the following:

```rust
trait Pod: Copy { }
```

That is,

* Pod types can always safely be copied,
* Pod types don't have to be unsized.

Both statements are true (given the definition of Pod used in lrs, unrelated to the Pod mentioned in the linked issue.) But I had to remove the `Copy` bound because I couldn't implement

```rust
impl<T: Pod> Pod for [T]
```

which is necessary because I want to implement

```rust
impl<T: Pod+?Sized> AsMut<[d8]> for T
```

The comment by @nikomatsakis seems to mostly address the cases where you're using `Copy+?Sized` directly, that is, where you care about `Copy`. But in the `AsRef` implementation I only care about the `Pod` aspect.

It seems that the built-in `Sized` bound can be removed in a backwards compatible way by adding an implicit `Sized` bound in libcore (should already exist via the `Clone` bound). If you then allowed `Copy` to be implemented for slices, the code above would work.

**Comment by cyplo** [comment]
Any news on this one? Do we need RFC for it ?

**Comment by nikomatsakis** [comment]
I am not sure whether I think we should permit sized types to be copy. I outlined [some of my thoughts in this comment](https://github.com/rust-lang/rust/issues/25776#issuecomment-106393457), which @eddyb already linked to. However, in the meantime, I've started to wonder if we should support moving and copying of DSTs just by using allocas at codegen time, which would allow us to make them work much more uniformly, almost like any other type. That would also make it fine for unsized types to be copy, though it might lead to footguns in some cases, but no different than a type like `[u32; 1024]` being `Copy`).
